be/src/io/cache_input_stream.cpp

```diff
@@ -142,42 +144,91 @@ Status CacheInputStream::_read_block(int64_t offset, int64_t size, char* out, bo
     }
     if (!res.is_not_found() && !res.is_resource_busy()) return res;
 
-    // read remote
-    char* src = nullptr;
     if (sb) {
         // Duplicate the block ranges to avoid saving the same data both in cache and shared buffer.
         _deduplicate_shared_buffer(sb);
-        const uint8_t* buffer = nullptr;
-        RETURN_IF_ERROR(_sb_stream->get_bytes(&buffer, block_offset, load_size));
-        strings::memcpy_inlined(out, buffer + shift, size);
-        src = (char*)buffer;
-    } else {
-        if (!can_zero_copy || (shift != 0)) {
-            can_zero_copy = false;
-            src = _buffer.data();
+    }
+
+    return Status::NotFound("Not Found");
+}
+
+Status CacheInputStream::_read_blocks_from_remote(const int64_t offset, const int64_t size, char* out) {
+    const int64_t start_block_id = offset / _block_size;
+    const int64_t end_block_id = (offset + size - 1) / _block_size;
+
+    // We will load range=[read_start_offset, read_end_offset) from remote
+    const int64_t block_start_offset = start_block_id * _block_size;
+    const int64_t block_end_offset = std::min(end_block_id * _block_size + _block_size, _size);
+
+    // cursors for `out`
+    int64_t out_offset_cursor = offset;
+    int64_t out_remain_size = size;
+    char* out_pointer_cursor = out;
+
+    for (int64_t read_offset_cursor = block_start_offset; read_offset_cursor < block_end_offset;) {
```

Comment:
I think the variables looks like `read_offset_cursor`, `out_offset_cursor` are too lengthy and make the code not legible.

Reply:
You can think about a better name in your next pr